src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs

```diff
@@ -65,30 +65,31 @@ internal static class LogRecordExtensions
         internal static OtlpLogs.LogRecord ToOtlpLog(this LogRecord logRecord)
         {
             OtlpLogs.LogRecord otlpLogRecord = null;
+            ref LogRecordData data = ref logRecord.GetDataRef();
```

Comment:
Just confirming my own understanding of the improvements this PR brings... so this is more performant because accessing properties of the struct results in a `call` vs. `callvirt`, right?

Reply:
This particular change is more to prevent a perf regression! This method used to work something like...

```csharp
ldarg.1 -> call LogRecord.CategoryName.Get() -> ldfld CategoryName_Backing -> ret
```

(Because `LogRecord` is sealed it should already be a `call` instead of `callvirt`.)

Now due to the pooling changes it has to do...

```csharp
ldarg.1 -> call LogRecord.CategoryName.Get() -> ldflda this.logRecordData -> call LogRecordData.CategoryName.Get() -> ldfld CategoryName_Backing -> ret
```

Extra call!

Using `GetDataRef()` it goes back to a single call like...

```csharp
ldflda [address of ref] -> call LogRecordData.CategoryName.Get()-> ldfld CategoryName_Backing -> ret
```

Just an optimization to cut out the extra hops.

True perf really comes down to how the JIT decides to inline things. Here's a [sharplab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0ASgK4B2wAllAKb4DCA9lAA4UA21ATgMpcBuFAMbUIAbnTo4AZlwJsdbAG902VbhlwALNgCySABQAZBgHNC1QQ04ATbG1PnLNgJRKVaj3hyCwwaiasATwA5MBpsAF47Bwsra3pffyDQmnE0D08kHBoICDATakjos1ibfB0RPIK0jLV3WrwATn0fPwDOELDqZxra3CRmnKru3tUAX3rsSelcbR0EIxinW3sS5ddldL7OagAzbGM1uIARXzBsazOinf3Vxzj8O9LrU+AwUYzJzO9E9s7wqKXN4JNrJLofb7YIb5QqAs7lSowiF1LYNAYtX5gmg9L5qJr6aEFHp9Kao7BjDyTaYyCDATjEQTAA5LE5nSabEkzLzyTEdFKFRSkkkeADaFWAAAsGNYAJIsNj6cVS2XygDyzEoDFIEHwwQYMtIbAopGNJmcAF1cbUCsBRELhaoINRbeSrep+jgKrkYUp7Q6xc7lXLmAqldLg2x1Zrtbr9YbjaaLW6PDa7cm1E6XRMydnKWSuXJDvcbOy3ca/JxSGA2MyjjZXucnssG8jVG6uVkeaC+V1fenVAHJeH5YrA8OQ1GKFqdXqDUaTaQzZaySSbZEAHzFYsveF0Xn/aitjKZjdb54NkFJHsA7B8avEQ+uld4jSdr3DPvP2qDoMjsMqicNSnGNZ3jBcl37bA1wiTcm1ZYF3yRP1hRPGCz2beFEIKIo7zYB87VzNRszGIA==) if you want to mess with it.

Reply:
So what are the real perf gains on this PR?

I will attempt to answer this 😄 

Our log logic was like this:

```csharp
                var record = new LogRecord(
                    provider.IncludeScopes ? this.ScopeProvider : null,
                    DateTime.UtcNow,
                    this.categoryName,
                    logLevel,
                    eventId,
                    provider.IncludeFormattedMessage ? formatter?.Invoke(state, exception) : null,
                    provider.ParseStateValues ? null : state,
                    exception,
                    provider.ParseStateValues ? this.ParseState(state) : null);

                processor.OnEnd(record);

                record.ScopeProvider = null;
```

And is now like this:

```csharp
                var record = LogRecordPool.Rent();

                record.ScopeProvider = provider.IncludeScopes ? this.ScopeProvider : null;
                record.State = provider.ParseStateValues ? null : state;
                record.StateValues = provider.ParseStateValues ? this.ParseState(record, state) : null;

                ref LogRecordData data = ref record.Data;

                data.TimestampBacking = DateTime.UtcNow;
                data.CategoryName = this.categoryName;
                data.LogLevel = logLevel;
                data.EventId = eventId;
                data.Message = provider.IncludeFormattedMessage ? formatter?.Invoke(state, exception) : null;
                data.Exception = exception;

                LogRecordData.SetActivityContext(ref data, Activity.Current);

                processor.OnEnd(record);

                record.ScopeProvider = null;

                // Attempt to return the LogRecord to the pool. This will no-op
                // if a batch exporter has added a reference.
                LogRecordPool.Return(record);
```

Should be slower, right?

Turns out, is only slightly slower (when hitting the `[ThreadStatic]`):

|       Method |     Mean |    Error |   StdDev |  Gen 0 | Allocated |
|------------- |---------:|---------:|---------:|-------:|----------:|
| LogUsingCtor | 38.89 ns | 0.154 ns | 0.137 ns | 0.0102 |     128 B |
| LogUsingPool | 40.37 ns | 0.054 ns | 0.048 ns |      - |         - |

But we eliminated all the memory pressure. I think this is why the stress test is performing better with the pool. Only a few more CPU cycles to use the pool but the GC is asleep freeing up the CPU to process more logs. This is a largely a guess but kind of makes sense?

Now that benchmark is interesting. How is it introducing the pool logic made it only slightly slower than just calling ctor?

Check out this benchmark:

```csharp
    public class CtorBenchmarks
    {
        [Benchmark]
        public TestSmallClass CtorSmallClass()
        {
            return new();
        }

        [Benchmark]
        public TestLargeClass CtorLargeClass()
        {
            return new();
        }

        public class TestSmallClass
        {
            public string? Prop1;
        }

        public class TestLargeClass
        {
            public string? Prop1;
            public string? Prop2;
            public string? Prop3;
            public int Prop4;
            public int Prop5;
            public int Prop6;
            public DateTime Prop7;
            public object? Prop8;
            public object? Prop9;
            public object? Prop10;
            public string? Prop11;
            public string? Prop12;
            public string? Prop13;
        }
    }
```

|         Method |     Mean |     Error |    StdDev |  Gen 0 | Allocated |
|--------------- |---------:|----------:|----------:|-------:|----------:|
| CtorSmallClass | 1.884 ns | 0.0236 ns | 0.0209 ns | 0.0019 |      24 B |
| CtorLargeClass | 3.982 ns | 0.0298 ns | 0.0279 ns | 0.0089 |     112 B |

Number of fields on the class impacts the ctor time. `LogRecord` has a lot of fields so that gives us some wiggle room to execute our pool logic before just calling the ctor would be faster. Interesting, eh?